forgerock-openbanking-analytics-core/src/main/java/com/forgerock/openbanking/analytics/upgrade/UpgradeStep_4_1_4.java

```diff
@@ -0,0 +1,108 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.forgerock.openbanking.analytics.upgrade;
+
+import com.forgerock.openbanking.analytics.model.entries.*;
+import com.forgerock.openbanking.analytics.models.TokenUsageEntry;
+import com.forgerock.openbanking.upgrade.exceptions.UpgradeException;
+import com.forgerock.openbanking.upgrade.model.UpgradeMeta;
+import com.forgerock.openbanking.upgrade.model.UpgradeStep;
+import lombok.extern.slf4j.Slf4j;
+import org.joda.time.DateTime;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.mongodb.core.BulkOperations;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.query.Criteria;
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.data.mongodb.core.query.Update;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StopWatch;
+
+@Service
+@UpgradeMeta(version = "4.1.4")
+@Slf4j
+public class UpgradeStep_4_1_4 implements UpgradeStep {
+
+    @Autowired
+    private MongoTemplate mongoTemplate;
+
+    @Override
+    public boolean upgrade() throws UpgradeException {
+        StopWatch elapsedTime = new StopWatch();
+        elapsedTime.start();
+        log.info("-----------------------------------------------------------------------");
+        log.debug("Start "+this.getClass().getName()+" to upgrading to version " + this.getClass().getDeclaredAnnotation(UpgradeMeta.class).version());
+        try {
+            upgradeEntries(TppEntry.class);
+            upgradeEntries(ConsentStatusEntry.class);
+            upgradeEntries(DirectoryCounterEntry.class);
+            upgradeEntries(EndpointUsageAggregate.class);
+            upgradeEntries(JwtsGenerationEntry.class);
+            upgradeEntries(JwtsValidationEntry.class);
+            upgradeEntries(PsuCounterEntry.class);
+            upgradeEntries(SessionCounterEntry.class);
+            upgradeEntries(TokenUsageEntry.class);
+            elapsedTime.stop();
+            log.info("Upgraded executed in " + elapsedTime.getTotalTimeSeconds() + " seconds.");
+            log.info("-----------------------------------------------------------------------");
+        } catch (Exception e) {
+            throw new UpgradeException("Could not upgrade to " + this.getClass().getDeclaredAnnotation(UpgradeMeta.class).version(), e);
+        } finally {
+            SecurityContextHolder.clearContext();
+        }
+        return true;
+    }
+
+    private void upgradeEntries(Class clazz) throws Exception {
```

Comment:
Minor - out of interest, do you see warnings in IntelliJ - e.g. like the use of a raw Class with no parameterized type, or the fact that it doesn't throw Exception? I can't ignore them! 😆 

You can change this methods to:

```
private <T> void upgradeEntries(Class<T> clazz) {
```

(Plus the methods below).

Reply:
I don't have any warning but done.